Implementation of parent and plugin overriding

davidfestal opened the pull request:
### What does this PR do?

This PR implements the parent or plugin overriding.

The overriding of parent devfile or devfile plugins is implemented using the Strategic merge patch K8S library.

#### Implementation entry points;

The implementation is available for import in the [`pkg/utils/overriding`](https://github.com/devfile/kubernetes-api/blob/overriding-implementation/pkg/utils/overriding) package.

Available methods are:
- [`OverrideDevWorkspaceTemplateSpec()`](https://github.com/devfile/kubernetes-api/blob/overriding-implementation/pkg/utils/overriding/overriding.go#L200) to work with GO structs directly
- [`OverrideDevWorkspaceTemplateSpecBytes`](https://github.com/devfile/kubernetes-api/blob/overriding-implementation/pkg/utils/overriding/overriding.go#L106) to work with `json` or `yaml` content.

In both case the result is a `DevWorkspaceTemplateSpecContent` Go struct that corresponds to the code content of a devfile 2.0 (without the top-level `apiVersion` and `metadata` fields).

Note that one can only override an **existing** element of a parent or a plugin.

If a new element is added, it should be added in the main (child) devfile body. And then simple merging of top-level lists is done between the main devfile body, the flattened parent and the flattened plugins. If the key (name of ID) of a element is found several times in the merge, an error is thrown.

The corresponding methods are:
- [`MergeDevWorkspaceTemplateSpec()`](https://github.com/devfile/api/blob/overriding-implementation/pkg/utils/overriding/merging.go#L208) to work with GO structs directly
- [`MergeDevWorkspaceTemplateSpecBytes`](https://github.com/devfile/api/blob/overriding-implementation/pkg/utils/overriding/merging.go#L154) to work with `json` or `yaml` content.

In both case the result is a `DevWorkspaceTemplateSpecContent` Go struct that corresponds to the code content of a devfile 2.0 (without the top-level `apiVersion` and `metadata` fields).

#### Examples of use:

- An example of use with GO structs can be found in the following test: https://github.com/devfile/kubernetes-api/blob/overriding-implementation/pkg/utils/overriding/overriding_test.go#L160
- An example of use with YAML files structs can be found in the following test: https://github.com/devfile/kubernetes-api/blob/overriding-implementation/pkg/utils/overriding/overriding_test.go#L168
- A similar example exists for the merging part.

#### YAML examples:

YAML example files are provided and used in GO tests.

##### Overriding

Overriding examples can be found in the following [`test-fixtures/patches`](https://github.com/devfile/kubernetes-api/tree/overriding-implementation/pkg/utils/overriding/test-fixtures/patches) folder: each sub-folder contains 3 Yaml files:
- `original.yaml`: the original Yaml that should be overridden  
- `patch.yaml`: the Yaml patch that should be used to override the original file
- `result.yaml`: the expected result of the overriding, which is the result of applying strategic merge patch on the original file.  

##### Merging

Merging examples can be found in the following [`test-fixtures/merges`](https://github.com/devfile/kubernetes-api/tree/overriding-implementation/pkg/utils/overriding/test-fixtures/merges) folder: each sub-folder contains up to 4 Yaml files:
- `main.yaml`: the Yaml of the main (child) devfile body content     
- `parent.yaml`: the Yaml of a flattened parent referenced by the main devfile 
- `plugin.yaml`: the Yaml a flattened plugin referenced by the main devfile
- `result.yaml`: the expected result of the merging of all those devfile contents.  

When opened in VSCode or [`Hosted Che`](https://che.openshift.io/f/?url=https://github.com/devfile/kubernetes-api/tree/overriding-implementation), all these example files are validated with appropriate and up-to-date schemas. So they benefit from YAML language services such as completion and doc hovers. 

Contributions are welcomed in the future to further complete the tests (and at the same time documentation) of overriding rules.

#### Noteworthy (and possibly breaking) changes

- [`Events`](https://github.com/devfile/kubernetes-api/pull/98/files?file-filters%5B%5D=.go#diff-80b8e2cc67c3cc3fc4cd18ef906b1316R22):
  - Detailed change: Instead of directly referencing the `Events` structure, the `DevWorkspaceTemplateSpecContent` struct contains a pointer to an `Events` struct.
  - Reason: This make testing easier, since it avoids a number of corner cases around serialization, when an empty `Events` object is created although no `events` field is present in the parsed `json` or `yaml`.
  - Impact: it will require a change in depending code, and a nil-check after json or yaml parsing.

#### Implementation details

##### Keys <span id="keys"></span>

Currently in the existing model, for some top-level lists (`components`, `commands`), the keys of elements are not directly in the objects stored in the list, but one level lower:

```yaml
commands:
  - exec:
      id: "commandId"  <------ the key of the given command in the list of commands 
```
instead of 
```yaml
commands:
  - id: "commandId"
    exec:
      ....
```

Furthermore for some components as `plugin`s, the key of a plugin in the list of components may be calculated, and may depend on a number of fields (`id` and `registryUrl` for example).
Thus a [`Keyed`](https://github.com/devfile/kubernetes-api/blob/overriding-implementation/pkg/apis/workspaces/v1alpha1/keyed.go#L3) interface is defined and implemented for commands, components and component overrides, and it is then used to setup fixed keys in the `commands` and `components` lists of both the original object and the patch object during overriding.

##### Union normalizing

In the GO API, some model types are defined as [Kubernetes unions](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190325-unions.md) with discriminators.

The API [`Union`](https://github.com/devfile/kubernetes-api/blob/overriding-implementation/pkg/apis/workspaces/v1alpha1/union.go#L6) interface defines helpers to manage correctly manage those unions, especially maintaining the fact that values in a union are mutually-exclusive.

This union management logic is then [used in the overriding implementation](https://github.com/devfile/kubernetes-api/blob/overriding-implementation/pkg/utils/overriding/overriding.go#L147), but is also available in a dedicated [`pkg/utils/unions`](https://github.com/devfile/kubernetes-api/tree/overriding-implementation/pkg/utils/unions) package that allows managing all the unions embedded in a wider object.

### What issues does this PR fix or reference?

This PR references issue https://github.com/devfile/kubernetes-api/issues/91

### Is your PR tested? Consider putting some instruction how to test your changes

The overriding logic implemented in this PR has associated tests

kadel commented:
Going over this PR I noticed something that for some reason I didn't notice before.

The current design looks quite weird, and it is not how Strategic Merge Patch works in Kubernetes.

https://github.com/devfile/kubernetes-api/issues/25#issuecomment-661783929

davidfestal commented:
> Going over this PR I noticed something that for some reason I didn't notice before.
> 
> The current design looks quite weird, and it is not how Strategic Merge Patch works in Kubernetes.
> 
> More in my comment at [#25 (comment)](https://github.com/devfile/api/issues/25#issuecomment-661783929)

@kadel I implemented the agreement we found (summarized [here](https://github.com/devfile/api/issues/25#issuecomment-668672087) ) in the PR, and updated the PR description to add details about the *merging* side of things (addition of elements that are not present in parents or plugins).

davidfestal commented:
cc @ranakan19

amisevsk commented on pkg/utils/overriding/merging.go:
```diff
@@ -0,0 +1,229 @@
+	return checkKeys(func(elementType string, keysSets []sets.String) []error {
+		mainKeys := keysSets[0]
+		parentOrPluginKeys := keysSets[1]
+		overridenElementsInMainContent := mainKeys.Intersection(parentOrPluginKeys)
```
```suggestion
		overriddenElementsInMainContent := mainKeys.Intersection(parentOrPluginKeys)
```

amisevsk commented on pkg/utils/overriding/merging.go:
```diff
@@ -0,0 +1,229 @@
+	allContents = append(allContents, pluginFlattenedContents...)
+	allContents = append(allContents, mainContent)
+	for _, content := range allContents {
+		addKeys(content)
```
Ignored error

amisevsk commented on pkg/utils/overriding/overriding.go:
```diff
@@ -0,0 +1,157 @@
+package overriding
+
+import (
+	//	"errors"
```
```suggestion
```

amisevsk commented on pkg/utils/overriding/overriding.go:
```diff
@@ -0,0 +1,157 @@
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+func ensureOnlyExistingElementsAreOverriden(keyedSpec *keyedDevWorkspaceTemplateSpecContent, keyedOverlay *keyedDevWorkspaceTemplateSpecContent) error {
```
```suggestion
func ensureOnlyExistingElementsAreOverridden(keyedSpec *keyedDevWorkspaceTemplateSpecContent, keyedOverlay *keyedDevWorkspaceTemplateSpecContent) error {
```

amisevsk commented:
I had to rebase on master as there was a conflict. Commit https://github.com/devfile/api/pull/98/commits/13fbbc9d6b9f9e5c19feee1678885e83492b700a resolves the minor comments from my review.

davidfestal commented:
Thanks @amisevsk for this deep review and very-valuable fixes.
